### PR TITLE
[After #3716] Fix debug assert failures on some tests for parallel discovery

### PR DESCRIPTION
**Pull request by Evangelink**
Not sure why we missed these cases but some tests are failing in debug mode because of some `Debug.Assert` conditions.

### CODE REVIEW

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyDiscoveryManager.cs [hunk 1]
```diff
@@ -27,7 +27,7 @@ internal class ParallelProxyDiscoveryManager : IParallelProxyDiscoveryManager
     private readonly ParallelOperationManager<IProxyDiscoveryManager, ITestDiscoveryEventsHandler2, DiscoveryCriteria> _parallelOperationManager;
     private readonly Dictionary<string, TestRuntimeProviderInfo> _sourceToTestHostProviderMap;
     private int _discoveryCompletedClients;
-    private int _availableTestSources = -1;
```
Comment (Evangelink) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyDiscoveryManager.cs:30]: It brings nothing to initialize with `-1`.

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Client/Parallel/ParallelProxyDiscoveryManagerTests.cs [hunk 2]
```diff
@@ -147,7 +147,7 @@ public void DiscoverTestsShouldProcessAllSources()
     }
 
     [TestMethod]
-    public void HandlePartialDiscoveryCompleteShouldReturnTrueIfDiscoveryWasAbortedWithEventHandler()
+    public void HandlePartialDiscoveryCompleteShouldReturnTrueIfDiscoveryWasAbortedBeforeBeingStartedWithEventHandler()
```
Comment (Evangelink) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Client/Parallel/ParallelProxyDiscoveryManagerTests.cs:150]: This test calls `Abort` without having called `DiscoverTests` before, that seems illogical but nothing in our code prevents this state so I decided to keep the test for now.

As we were initializing to -1 the number of sources and never calling `DiscoverTests` which resets the value to the actual number of sources (i.e. 0+) we were having assertion failures.

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Client/Parallel/ParallelProxyDiscoveryManagerTests.cs [hunk 3]
```diff
@@ -159,7 +159,22 @@ public void HandlePartialDiscoveryCompleteShouldReturnTrueIfDiscoveryWasAbortedW
...
+
+        parallelDiscoveryManager.DiscoverTests(_discoveryCriteriaWith2Sources, _mockEventHandler.Object);
+        parallelDiscoveryManager.Abort(_mockEventHandler.Object);
+        bool isPartialDiscoveryComplete = parallelDiscoveryManager.HandlePartialDiscoveryComplete(proxyDiscovermanager, 20, new List<TestCase>(), isAborted: false);
+
+        Assert.IsTrue(isPartialDiscoveryComplete);
+    }
+
+    [TestMethod]
+    public void HandlePartialDiscoveryCompleteShouldReturnTrueIfDiscoveryWasAbortedBeforeBeingStarted()
```
Comment (Evangelink) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Client/Parallel/ParallelProxyDiscoveryManagerTests.cs:177]: Similar to comment test aborting without starting discovery.

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Client/Parallel/ParallelProxyDiscoveryManagerTests.cs [hunk 4]
```diff
@@ -170,13 +185,28 @@ public void HandlePartialDiscoveryCompleteShouldReturnTrueIfDiscoveryWasAborted(
         Assert.IsTrue(isPartialDiscoveryComplete);
     }
 
+    [TestMethod]
+    public void HandlePartialDiscoveryCompleteShouldReturnTrueIfDiscoveryWasAbortedAfterBeingStarted()
```
Comment (Evangelink) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Client/Parallel/ParallelProxyDiscoveryManagerTests.cs:189]: Similar to comment test aborting after starting discovery.

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Client/Parallel/ParallelProxyDiscoveryManagerTests.cs (hunk 2 above)
Comment (nohwnd) [reply test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Client/Parallel/ParallelProxyDiscoveryManagerTests.cs:150]: Cool, I planned to investigate why I am seeing the failure on -1 sources, I thought it is something like this where the initial step is skipped in a unit test. So now I don't have to.

**Pull request merged**